runtime/src/main/java/org/capnproto/BuilderArena.java

```diff
@@ -76,32 +76,19 @@ public AllocateResult(SegmentBuilder segment, int offset) {
     }
 
     public AllocateResult allocate(int amount) {
-
         int len = this.segments.size();
-        // we allocate the first segment in the constructor.
 
-        int result = this.segments.get(len - 1).allocate(amount);
-        if (result != SegmentBuilder.FAILED_ALLOCATION) {
-            return new AllocateResult(this.segments.get(len - 1), result);
+        // we allocate the first segment in the constructor.
+        if (len > 0) {
```

Comment:
shouldn't here be a test that segments are still <=512?

Reply:
Hmm, maybe!

We do check for that on the reading side: https://github.com/capnproto/capnproto-java/blob/45fd638c7cc7297098ea7a9b72e77e695fd45cff/runtime/src/main/java/org/capnproto/Serialize.java#L67

 I think the main reason for that test is to protect against resource exhaustion attacks. It seems less urgent to have a similar check on the writing size (but, still, maybe worthwhile). As far as I know, it's not actually part of the encoding spec.

Reply:
the problem is that you can easily generate messages, that nobody can read (the capn Proto decoder binary also complains about 512 segments). a similar thing happens with the traversal limit (basically limiting messages to 8 Mio words for Java.)